pkg/controller/infinispan/resources/service/service_controller.go

```diff
@@ -141,26 +140,82 @@ func (s serviceResource) Process() (reconcile.Result, error) {
-		// Set Infinispan instance as the owner and controller
-		controllerutil.SetControllerReference(s.infinispan, resource, s.scheme)
-		err := s.client.Create(ctx, ro)
-		if err != nil {
-			s.log.Error(err, fmt.Sprintf("failed to create %s", kind), kind, resource)
-			return err
+func (s serviceResource) reconcileService(service *corev1.Service) error {
+	findService := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      service.GetName(),
+			Namespace: service.GetNamespace(),
+		},
+	}
+	result, err := controllerutil.CreateOrUpdate(context.TODO(), s.client, findService, func() error {
+		if findService.CreationTimestamp.IsZero() {
+			controllerutil.SetControllerReference(s.infinispan, findService, s.scheme)
+			findService.Spec = service.Spec
+			findService.Annotations = service.Annotations
+		}
+		findService.Labels = service.Labels
+		return nil
+	})
+	if err != nil {
+		s.log.Error(err, "failed to create or update Service", service)
+		return err
+	}
+	if result != controllerutil.OperationResultNone {
+		s.log.Info(fmt.Sprintf("%s Service %s", strings.Title(string(result)), findService.GetName()))
+	}
+	return nil
+}
+
+func (s serviceResource) reconcileRoute(route *routev1.Route) error {
+	findRoute := &routev1.Route{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      route.GetName(),
+			Namespace: route.GetNamespace(),
+		},
+	}
+	result, err := controllerutil.CreateOrUpdate(context.TODO(), s.client, findRoute, func() error {
+		if findRoute.CreationTimestamp.IsZero() {
+			controllerutil.SetControllerReference(s.infinispan, findRoute, s.scheme)
+			findRoute.Spec = route.Spec
+			findRoute.Annotations = route.Annotations
 		}
-		s.log.Info(fmt.Sprintf("Created %s", kind), kind, resource.GetName())
+		findRoute.Labels = route.Labels
 		return nil
+	})
+	if err != nil {
+		s.log.Error(err, "failed to create or update Route", route)
+		return err
 	}
-	return err
+	if result != controllerutil.OperationResultNone {
+		s.log.Info(fmt.Sprintf("%s Route %s", strings.Title(string(result)), findRoute.GetName()))
+	}
+	return nil
+}
+
+func (s serviceResource) reconcileIngress(ingress *networkingv1beta1.Ingress) error {
+	findIngress := &networkingv1beta1.Ingress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ingress.GetName(),
+			Namespace: ingress.GetNamespace(),
+		},
+	}
+	result, err := controllerutil.CreateOrUpdate(context.TODO(), s.client, findIngress, func() error {
+		if findIngress.CreationTimestamp.IsZero() {
+			controllerutil.SetControllerReference(s.infinispan, findIngress, s.scheme)
+			findIngress.Spec = ingress.Spec
+			findIngress.Annotations = ingress.Annotations
+		}
+		findIngress.Labels = ingress.Labels
+		return nil
+	})
+	if err != nil {
+		s.log.Error(err, "failed to create or update Ingress", ingress)
+		return err
+	}
+	if result != controllerutil.OperationResultNone {
+		s.log.Info(fmt.Sprintf("%s Ingress %s", strings.Title(string(result)), findIngress.GetName()))
+	}
+	return nil
```

Comment:
AFAICT all three reconcile functions are identical, except the object type being updated. Could we use `ToUnstructured` to create a generic method to handle the following part of the code:

```
		result, err := controllerutil.CreateOrUpdate(context.TODO(), s.client, findIngress, func() error {
		if findIngress.CreationTimestamp.IsZero() {
			controllerutil.SetControllerReference(s.infinispan, findIngress, s.scheme)
			findIngress.Spec = ingress.Spec
			findIngress.Annotations = ingress.Annotations
		}
		findIngress.Labels = ingress.Labels
		return nil
	})
	if err != nil {
		s.log.Error(err, "failed to create or update Ingress", ingress)
		return err
	}
	if result != controllerutil.OperationResultNone {
		s.log.Info(fmt.Sprintf("%s Ingress %s", strings.Title(string(result)), findIngress.GetName()))
	}
```

Reply:
This could be an option, but I we would like to change some `Spec` structures not sure that will be a possible. For `ObjectMeta` it should be possible

Reply:
I haven't tried it, but I think you could do something like:

```go
		found := unstructured.ToUnstructured(findIngress)
		provided := unstructured.ToUnstructured(ingress)
		unstructured.SetNestedField(found, provided["Spec"], "Spec")
		unstructured.SetNestedField(found, provided["Annotations"], "Annotations")
```
But maybe the saved lines aren't worth it :slightly_smiling_face:

Reply:
Thanks for idea, @ryanemerson 
Bit polished and used for common function. But we should avoid to use `Unstructured` a lot.

Reply:
Fixed, but Baclup/Restore test failed with another error